actors/evm/src/interpreter/precompiles.rs

```diff
@@ -57,79 +78,145 @@ impl<BS: Blockstore, RT: Runtime<BS>> Precompiles<BS, RT> {
     }
 }
 
-/// read 32 bytes (u256) from buffer or error
-fn read_u256(buf: &[u8], start: usize) -> Result<U256, PrecompileError> {
-    let slice = buf.get(start..start + 32).ok_or(PrecompileError::IncorrectInputSize)?;
-    Ok(U256::from_big_endian(slice))
+/// Intermediary container to hold a data as slice or turn into a vec with padding as needed.
+/// This is intended to not copy in normal operations
+enum Data<'a> {
+    Slice(&'a [u8]),
+    Vec(Vec<u8>),
 }
+impl<'a> Data<'a> {
+    fn slice(&self) -> &[u8] {
+        match self {
+            Self::Slice(s) => s,
+            Self::Vec(v) => v,
+        }
+    }
+
+    /// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/common/bytes.go#L108
+    fn read_right_pad(input: &'a [u8], len: usize) -> Self {
+        if len <= input.len() {
+            Self::Slice(input)
+        } else {
+            let mut padded = vec![0u8; len];
+            padded[..input.len()].copy_from_slice(input);
```

Comment:
```suggestion
            let mut padded = Vec::with_capacity(len);
            padded.extend(input);
            padded.resize(len - input.len(), 0);
```

Less zero initialization.